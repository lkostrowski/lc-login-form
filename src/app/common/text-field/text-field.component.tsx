import React, { HTMLProps, ReactNode, useRef } from 'react';
import cx from 'classnames';
import nanoID from 'nanoid';

import styles from './text-field.module.scss';

export interface Props extends HTMLProps<HTMLInputElement> {
    label?: string;
    inputClassName?: string;
    errorMessage?: string;
    hasError?: boolean;
    rightIcon?: ReactNode;
}

export const TextField: React.FC<Props> = ({
    label,
    className,
    inputClassName,
    errorMessage,
    hasError,
    rightIcon,
    ...props
}) => {
    const id = useRef(nanoID(5));

    return (
        <div className={cx(styles.container, className)}>
            {label && (
                <label
                    className={cx(styles.label, {
                        [styles.withError]: hasError,
                    })}
                    htmlFor={id.current}
                >
                    {label}
                </label>
            )}
            <div className={styles.fieldContainer}>
                <input
                    className={cx(inputClassName, styles.field, {
                        [styles.withError]: hasError,
                        [styles.withRightIcon]: Boolean(rightIcon),
                    })}
                    id={id.current}
                    {...props}
                />
                {rightIcon && (
                    <div className={styles.rightIconContainer}>{rightIcon}</div>
                )}
            </div>
            {errorMessage && hasError && (
                <span
                    data-testid="text-field:error-message"
                    className={styles.error}
                >
                    {errorMessage}
                </span>
            )}
        </div>
    );
};
