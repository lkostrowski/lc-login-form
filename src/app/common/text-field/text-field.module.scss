@import '../../../theme/colors';
@import '../../../theme/vars';

.container {
}

.field {
    background: $lighterGray;
    height: 5rem;
    border-radius: $borderRadius;
    box-shadow: none;
    outline: none;
    display: block;
    width: 100%;
    padding: 0 1rem;
    font-size: 1.4rem;
    border: 1px solid $lighterGray;

    &.withError {
        border-color: $red;
        animation: shake 0.6s;
    }

    &.withRightIcon {
        padding-right: 5rem;
    }
}

.fieldContainer {
    position: relative;
}

.label {
    margin-bottom: 0.5rem;
    display: block;
    color: $lightGray;
    font-size: 1.2rem;

    &.withError {
        color: $red;
    }
}

.error {
    color: $red;
    display: block;
    margin-top: 0.5rem;
    font-size: 1.2rem;
}

.rightIconContainer {
    position: absolute;
    right: 2rem;
    top: 50%;
    transform: translateY(-50%);
}

@keyframes shake {
    10%, 90% {
        transform: translate3d(-1px, 0, 0);
    }

    20%, 80% {
        transform: translate3d(2px, 0, 0);
    }

    30%, 50%, 70% {
        transform: translate3d(-4px, 0, 0);
    }

    40%, 60% {
        transform: translate3d(4px, 0, 0);
    }
}
