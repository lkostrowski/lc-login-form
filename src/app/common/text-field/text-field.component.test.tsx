import React from 'react';
import { cleanup, render } from '@testing-library/react';
import { TextField } from './text-field.component';
import nanoID from 'nanoid';

afterEach(cleanup);

jest.mock('nanoid');
(nanoID as jest.Mock).mockImplementation(() => 'id');

describe('Text field component', () => {
    describe('Snapshots', () => {
        test('Empty', () => {
            const { container } = render(<TextField />);

            expect(container).toMatchSnapshot();
        });
        test('Filled', () => {
            const { container } = render(
                <TextField value="Filled" onChange={() => {}} />,
            );

            expect(container).toMatchSnapshot();
        });
        test('With right icon', () => {
            const { container } = render(<TextField rightIcon={<svg />} />);

            expect(container).toMatchSnapshot();
        });
        test('With error', () => {
            const { container } = render(
                <TextField errorMessage="Some error" hasError />,
            );

            expect(container).toMatchSnapshot();
        });
    });
});
