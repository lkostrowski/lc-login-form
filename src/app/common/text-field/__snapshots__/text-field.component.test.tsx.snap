// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text field component Snapshots Empty 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="fieldContainer"
    >
      <input
        class="field"
        id="id"
      />
    </div>
  </div>
</div>
`;

exports[`Text field component Snapshots Filled 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="fieldContainer"
    >
      <input
        class="field"
        id="id"
        value="Filled"
      />
    </div>
  </div>
</div>
`;

exports[`Text field component Snapshots With error 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="fieldContainer"
    >
      <input
        class="field withError"
        id="id"
      />
    </div>
    <span
      class="error"
      data-testid="text-field:error-message"
    >
      Some error
    </span>
  </div>
</div>
`;

exports[`Text field component Snapshots With right icon 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="fieldContainer"
    >
      <input
        class="field withRightIcon"
        id="id"
      />
      <div
        class="rightIconContainer"
      >
        <svg />
      </div>
    </div>
  </div>
</div>
`;
